File: 백준/Gold/5972. 택배 배송/택배 배송.java
import java.util.*;
import java.io.*;

class Main {
    static final BufferedReader BR = new BufferedReader(new InputStreamReader(System.in));
    
    int[] inputArray;
    int N, M;
    int[] dijkstra;
    Map<Integer, Integer>[] roadMap;
    
    public static void main(String[] args) throws IOException {
        Main main = new Main();
        main.init();
        main.solve();
    }
    
    int[] getInputArray() throws IOException {
        return Arrays.stream(BR.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
    }
    
    void init() throws IOException {
        inputArray = getInputArray();
        N = inputArray[0];
        M = inputArray[1];
        roadMap = new Map[N+1];
        dijkstra = new int[N+1];
        for (int i = 0; i <= N; i++) {
            roadMap[i] = new HashMap<>();
            dijkstra[i] = 50_000_000;
        }
        
        for (int i = 0; i < M; i++) {
            inputArray = getInputArray();
            roadMap[inputArray[0]].put(inputArray[1], Math.min(roadMap[inputArray[0]].getOrDefault(inputArray[1], 50_000_000), inputArray[2]));
            roadMap[inputArray[1]].put(inputArray[0], Math.min(roadMap[inputArray[1]].getOrDefault(inputArray[0], 50_000_000), inputArray[2]));

        }
        
    }
    
    void solve() {
        PriorityQueue<Node> heap = new PriorityQueue<>(Node::compare);
        heap.add(new Node(1, 0));
        while (!heap.isEmpty()) {
            Node node = heap.remove();
            if (dijkstra[node.pos] <= node.value) {
                continue;
            }
            dijkstra[node.pos] = node.value;
            for (Integer next : roadMap[node.pos].keySet()) {
                if (dijkstra[next] > node.value + roadMap[node.pos].get(next)) {
                    heap.add(new Node(next, node.value + roadMap[node.pos].get(next)));
                }
            }
        }
        System.out.println(dijkstra[N]);
    }
    
    class Node {
        int pos;
        int value;
        
        Node(int pos, int value) {
            this.pos = pos;
            this.value = value;
        }
        
        int compare(Node compare) {
            return Integer.compare(this.value, compare.value);
        }
    }
}

File: 백준/Gold/5972. 택배 배송/택배 배송.kt
package org
import kotlin.math.*
import java.util.*
var br = System.`in`.bufferedReader()

fun main() {
    initialize()
}

fun getInputArr(): IntArray {
    return br.readLine().split(" ").map { it.toInt() }.toIntArray()
}

fun initialize() {
    val farmSize = getInputArr()
    val N: Int = farmSize[0]
    val M: Int = farmSize[1]

    val dijkstra = IntArray(N+1){50_000_000}
    val roadMap: Array<MutableMap<Int, Int>> = Array(N+1) {mutableMapOf()}
    for (i in 1..M) {
        val road = getInputArr()
        roadMap[road[0]][road[1]] = min(roadMap[road[0]][road[1]] ?: 50_000_000, road[2])
        roadMap[road[1]][road[0]] = min(roadMap[road[1]][road[0]] ?: 50_000_000, road[2])
    }
    solve(N, dijkstra, roadMap)
}

fun solve(N: Int, dijkstra: IntArray, roadMap: Array<MutableMap<Int, Int>>) {
    val heap:PriorityQueue<Node> = PriorityQueue<Node>(Node::compare)
    heap.add(Node(1, 0))
    while (!heap.isEmpty()) {
        val node = heap.remove()
        if (dijkstra[node.pos] <= node.value) {
            continue
        }
        dijkstra[node.pos] = node.value
        for ((k,v)  in roadMap[node.pos]) if (dijkstra[k] > node.value + v) {
            heap.add(Node(k, v + node.value))
        }
    }
    println(dijkstra[N])
}

class Node(val pos: Int, val value: Int) {
    fun compare(c: Node):Int {
        return this.value.compareTo(c.value)
    }
}